daemon/migrator.c

```diff
@@ -0,0 +1,805 @@
+            migrator_enter_state(self);
+
+            migrator_eval_state_later(self);
+        }
+    }
+}
+
+/* ------------------------------------------------------------------------- *
+ * MIGRATOR_QUEUE
+ * ------------------------------------------------------------------------- */
+
+static void
+migrator_build_queue(migrator_t *self)
+{
+    glob_t gl = {};
+
+    if( !glob(APPROVAL_PATTERN, 0, 0, &gl) ) {
+        for( int i = 0; i < gl.gl_pathc; ++i )
+            g_queue_push_tail(self->mgr_queue, g_strdup(gl.gl_pathv[i]));
+    }
+
+    globfree(&gl);
+}
+
+static bool
+migrator_queued(migrator_t *self)
+{
+    return self->mgr_queue->length != 0;
+}
+
+static gchar *
+migrator_dequeue(migrator_t *self)
+{
+    gchar *file = g_queue_pop_head(self->mgr_queue);
+    log_debug("migrator: dequeue: %s", file);
+    return file;
+}
+
+static void
+migrator_clear_queue(migrator_t *self)
+{
+    g_queue_clear_full(self->mgr_queue, g_free);
+}
+
+/* ------------------------------------------------------------------------- *
+ * MIGRATOR_REMOVAL_QUEUE
+ * ------------------------------------------------------------------------- */
+
+static void
+migrator_removal_enqueue(migrator_t *self, const gchar *path)
+{
+    g_queue_push_tail(self->mgr_removal_queue, g_strdup(path));
+}
+
+static bool
+migrator_removal_queued(migrator_t *self)
+{
+    return self->mgr_removal_queue->length != 0;
+}
+
+static void
+migrator_removal_dequeue_all(migrator_t *self)
+{
+    for( GList *iter = self->mgr_removal_queue->head; iter; iter = iter->next ) {
+        gchar *path = iter->data;
+        iter->data = NULL;
+        remove_approval_path(path);
+        g_free(path);
+    }
+    g_queue_clear_full(self->mgr_removal_queue, NULL);
+}
+
+static void
+migrator_removal_clear_queue(migrator_t *self)
+{
+    g_queue_clear_full(self->mgr_removal_queue, g_free);
+}
+
+/* ------------------------------------------------------------------------- *
+ * MIGRATOR_PROCESS
+ * ------------------------------------------------------------------------- */
+
+static void
+migrator_process_file(migrator_t *self, const gchar *path)
+{
+    bool           migrated    = false;
+    gchar         *appname     = NULL;
+    appsettings_t *appsettings = NULL;
+    approval_t    *approval    = approval_create(self, path);
+    stringset_t   *granted     = stringset_create();
+
+    if( approval_valid(approval) ) {
+        const stringset_t *available = control_available_applications(migrator_control(self));
+
+        appname = path_to_desktop_name(approval_profile(approval));
+        if( stringset_has_item(available, appname) ) {
+            const appinfo_t *appinfo = migrator_appinfo(self, appname);
+
+            if( !g_strcmp0(approval_organization(approval), appinfo_get_organization_name(appinfo)) &&
+                !g_strcmp0(approval_application(approval), appinfo_get_application_name(appinfo)) ) {
+                /* TODO: The plugin stored also binary path in the
```

Comment:
I'm not sure if we need to do this or not. It's also possible that if we can remove all instances of invoker and sailjail from Exec lines, then the existing check (marked with DEAD_CODE) would work as it is.

Reply:
IMO something like: the command line needs to be checked, but that would be something that the party that is going to handle actual execution needs to handle. Here in daemon side we allow/deny running applications, and execution details can and are expected to change without affecting allowed state -> we should ignore binary path in migration.